internal/cmd/gocdk/internal/static/static.go

```diff
@@ -14,12 +14,315 @@
 
 package static
 
-import "net/http"
-
-// Open opens a file from the static assets.
-// TODO(rvangent): Provide a richer (but hopefully smaller) API,
-// which should cover things like copying files from assets/,
-// materializeTemplateDir, etc.
-func Open(path string) (http.File, error) {
-	return assets.Open(path)
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path"
+	"path/filepath"
+	"text/template"
+
+	"golang.org/x/xerrors"
+)
+
+// ListFiles returns a listing of files from the static data under dir,
+// recursively. The relative path to each file from dir is returned.
+func ListFiles(dir string) ([]string, error) {
+	f, err := assets.Open(dir)
+	if err != nil {
+		return nil, err
+	}
+	infos, err := f.Readdir(-1)
+	if err != nil {
+		return nil, err
+	}
+	var retval []string
+	for _, info := range infos {
+		fullpath := path.Join(dir, info.Name())
+		if info.IsDir() {
+			nestedFiles, err := ListFiles(fullpath)
+			if err != nil {
+				return nil, err
+			}
+			for _, f := range nestedFiles {
+				retval = append(retval, path.Join(info.Name(), f))
+			}
+			continue
+		}
+		retval = append(retval, info.Name())
+	}
+	return retval, nil
+}
+
+// Action represents an action relative to static content that will be performed
+// in Do.
+type Action struct {
```

Comment:
Do we expect for someone to create an Action outside of using the functions you provide? It seems like the functions are a more approachable way of creating these object.

Reply:
Maybe? Not sure yet.

Reply:
I have a package that I wrote like this for gg: https://github.com/zombiezen/gg/blob/master/internal/filesystem/filesystem.go

I found it easier to keep the fields unexported and just use exported constructor functions to build the actions.

Reply:
That is definitely easier, and expected. I added a TODO to consider un-exporting the fields. At least `TemplateData` is used externally so would need to stay public.